### PR TITLE
[3.0][cdc-common] Introduce DataSource related APIs for Flink CDC 3.0

### LINKED ISSUE #2625: [flink-cdc-common] Introduce DataSource related APIs
### Search before asking

- [X] I searched in the [issues](https://github.com/ververica/flink-cdc-connectors/issues) and found nothing similar.

### Motivation

Intoruce DataSource related APIs

### Solution

_No response_

### Alternatives

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by Jiabao-Sun**
Introduce DataSource related APIs for Flink CDC 3.0.

This closes #2625

### CODE REVIEW

File: flink-cdc-common/src/main/java/com/ververica/cdc/common/source/FlinkSourceFunctionProvider.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+import com.ververica.cdc.common.event.Event;
+
+/**
+ * {@code FlinkSourceFunctionProvider} is used to provide a Flink {@link SourceFunction} for reading
+ * events from external systems.
+ */
+@PublicEvolving
+public interface FlinkSourceFunctionProvider extends EventSourceProvider {
+    SourceFunction<Event> getSourceFunction();
```
Comment (lvyanquan) [inline flink-cdc-common/src/main/java/com/ververica/cdc/common/source/FlinkSourceFunctionProvider.java:30]: We can provide a Helper method like [of](https://github.com/apache/flink/blob/fa8ea3f9a443f555de5c5479d725e16befefc957/flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/sink/SinkV2Provider.java#L40-L42C6)

**Review by lvyanquan — APPROVED** [review]

File: flink-cdc-common/src/main/java/com/ververica/cdc/common/source/DataSource.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * {@code DataSource} is used to access metadata and read change data from external systems. It can
+ * read data from multiple tables simultaneously.
+ *
+ * <p>The ability to perform full database synchronization comes from the implementation of the
+ * Flink source. After obtaining user configuration, the Flink source connector needs to be able to
+ * send all data from a specific database downstream according to the specified data type.
+ */
+@PublicEvolving
+public interface DataSource {
+    EventSourceProvider getEventSourceProvider();
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/com/ververica/cdc/common/source/DataSource.java:31]: minor: A blank line is recommended  before starting the interface body content

File: flink-cdc-common/src/main/java/com/ververica/cdc/common/source/DataSource.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    MetadataAccessor getMetadataAccessor();
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/com/ververica/cdc/common/source/DataSource.java:33]: Could we add java doc for public methods in public interface, these methods are part of Public API too.

File: flink-cdc-common/src/main/java/com/ververica/cdc/common/source/DataSource.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+import org.apache.flink.annotation.PublicEvolving;
+
+/**
+ * {@code DataSource} is used to access metadata and read change data from external systems. It can
+ * read data from multiple tables simultaneously.
+ *
+ * <p>The ability to perform full database synchronization comes from the implementation of the
+ * Flink source. After obtaining user configuration, the Flink source connector needs to be able to
+ * send all data from a specific database downstream according to the specified data type.
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/com/ververica/cdc/common/source/DataSource.java:27]: `entire database sync` is just one user case for the DataSource, we should focus on itself when explain the ability of the interface.

File: flink-cdc-common/src/main/java/com/ververica/cdc/common/source/MetadataAccessor.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import com.ververica.cdc.common.event.TableId;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+
+/** {@code MetadataAccessor} is used to read the metadata from external systems. */
+@PublicEvolving
+public interface MetadataAccessor {
+
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/com/ververica/cdc/common/source/MetadataAccessor.java:30]: The blank line looks better 😁

File: flink-cdc-common/src/main/java/com/ververica/cdc/common/source/MetadataAccessor.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import org.apache.flink.annotation.PublicEvolving;
+
+import com.ververica.cdc.common.event.TableId;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+
+/** {@code MetadataAccessor} is used to read the metadata from external systems. */
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/com/ververica/cdc/common/source/MetadataAccessor.java:27]:
```suggestion
/** {@code MetadataAccessor} is used by {@code DataSource} to access the metadata from external systems. */
```

File: flink-cdc-common/src/main/java/com/ververica/cdc/common/source/EventSourceProvider.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ververica.cdc.common.source;
+
+import org.apache.flink.annotation.PublicEvolving;
+
+/** A marker interface used to provide an event source for reading events from external systems. */
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/com/ververica/cdc/common/source/EventSourceProvider.java:21]: We can simply introducing current design: we can reuse exiting Flink Source and Flink SourceFunction implementation, and we can support our own EventSource Implementation in the future.

**Review by leonardBang — COMMENTED** [review]
Thanks @Jiabao-Sun for the contribution, I left some comments

**Review by leonardBang — APPROVED** [review]
Thanks @Jiabao-Sun for the update, LGTM

**Pull request merged**